File: backend/routers/audit.py
"""Audit trail and audit reports router."""
import uuid
from datetime import datetime, timezone
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import select, func
from sqlalchemy.ext.asyncio import AsyncSession

from middleware.tenant_rls import tenant_db_from_jwt, bypass_db
from middleware.auth import get_current_user, CurrentUser
from services.audit_logger import log_audit
from models.audit import AuditLog

router = APIRouter()

class AuditReportCreate(BaseModel):
    title: str
    period_start: datetime = None
    period_end: datetime = None

@router.get("/compliance/audit-trail")
async def list_audit_trail(
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
    limit: int = Query(50, ge=1, le=500),
    offset: int = Query(0, ge=0),
):
    """List audit log entries."""
    stmt = select(AuditLog).where(
        AuditLog.firm_id == current_user.firm_id
    ).order_by(AuditLog.created_at.desc()).limit(limit).offset(offset)
    result = await db.execute(stmt)
    logs = result.scalars().all()
    return [
        {
            "id": log.id,
            "firm_id": log.firm_id,
            "action": log.action,
            "entity_type": log.entity_type,
            "entity_id": log.entity_id,
            "user_id": log.user_id,
            "details": log.details,
            "ip_address": log.ip_address,
            "created_at": log.created_at,
        }
        for log in logs
    ]

@router.get("/compliance/audit-trail/summary")
async def audit_summary(
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Get audit summary statistics."""
    stmt = select(func.count(AuditLog.id)).where(
        AuditLog.firm_id == current_user.firm_id
    )
    result = await db.execute(stmt)
    total_actions = result.scalar() or 0
    
    stmt = select(func.count(func.distinct(AuditLog.user_id))).where(
        AuditLog.firm_id == current_user.firm_id
    )
    result = await db.execute(stmt)
    total_users = result.scalar() or 0
    
    stmt = select(AuditLog.action, func.count(AuditLog.id)).where(
        AuditLog.firm_id == current_user.firm_id
    ).group_by(AuditLog.action)
    result = await db.execute(stmt)
    action_counts = dict(result.all())
    
    return {
        "total_actions": total_actions,
        "total_users": total_users,
        "action_counts": action_counts,
    }

@router.get("/compliance/audit-reports")
async def list_audit_reports(
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """List audit reports."""
    return []

@router.post("/compliance/generate-audit-report")
async def generate_audit_report(
    payload: AuditReportCreate,
    current_user: CurrentUser = Depends(get_current_user),
    db: AsyncSession = Depends(tenant_db_from_jwt),
):
    """Generate audit report."""
    await log_audit(
        db=db,
        firm_id=current_user.firm_id,
        action="generated_report",
        entity_type="audit_report",
        entity_id=str(uuid.uuid4()),
        user_id=current_user.user_id,
        details=f"Generated audit report: {payload.title}",
    )
    return {
        "id": str(uuid.uuid4()),
        "title": payload.title,
        "period_start": payload.period_start,
        "period_end": payload.period_end,
        "generated_at": datetime.now(timezone.utc),
    }

File: archive/seema-node/src/routes/audit.ts
import { Router, Request, Response } from 'express';
import prisma from '../lib/prisma.js';
import { authenticate, requireRole } from '../middleware/auth.js';
import { getTenantFilter } from '../middleware/tenant.js';
import { logAudit } from '../middleware/auditLogger.js';
import { z } from 'zod';

const router = Router();

// Serialize Prisma's camelCase to the snake_case shape the frontend expects.
function serializeAuditLog(l: any) {
  return {
    id: l.id,
    user_id: l.userId,
    user_name: l.userName ?? null,
    action: l.action,
    entity_type: l.entityType,
    entity_id: l.entityId,
    ip_address: l.ipAddress,
    details: l.details ?? null,
    metadata: l.metadata,
    created_at: l.createdAt,
  };
}

function serializeScanResult(r: any) {
  return {
    id: r.id,
    scan_date: r.scanDate,
    category: r.category,
    check_name: r.checkName,
    status: r.status,
    details: r.details,
    recommendation: r.recommendation,
    created_at: r.createdAt,
    updated_at: r.updatedAt,
  };
}

// GET /compliance/audit-trail — List audit logs for firm
router.get('/compliance/audit-trail', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const logs = await prisma.auditLog.findMany({
      where: { firmId },
      orderBy: { createdAt: 'desc' },
      take: 100,
    });
    res.json(logs.map(serializeAuditLog));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch audit trail' });
  }
});

// GET /compliance/audit-trail/summary — Count by action type, grouped by entityType
router.get('/compliance/audit-trail/summary', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const summary = await prisma.auditLog.groupBy({
      by: ['entityType'],
      where: { firmId },
      _count: { action: true },
    });

    const formatted = summary.map((item) => ({
      entity_type: item.entityType,
      count: item._count.action,
    }));

    res.json(formatted);
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch audit summary' });
  }
});

// GET /compliance/audit-reports — List compliance scan results (report proxy)
router.get('/compliance/audit-reports', authenticate, async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);
    const reports = await prisma.complianceScanResult.findMany({
      where: { firmId },
      orderBy: { createdAt: 'desc' },
    });
    res.json(reports.map(serializeScanResult));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to fetch audit reports' });
  }
});

// POST /compliance/generate-audit-report — Create scan result as report placeholder
router.post('/compliance/generate-audit-report', authenticate, requireRole('admin'), async (req: Request, res: Response) => {
  try {
    const { firmId } = getTenantFilter(req);

    const report = await prisma.complianceScanResult.create({
      data: {
        firmId,
        scanDate: new Date(),
        category: req.body.category || 'audit_report',
        checkName: req.body.title || 'Generated Audit Report',
        status: 'completed',
        details: req.body.details || 'Audit report generated on demand',
        recommendation: req.body.recommendation || null,
      },
    });

    await logAudit({
      firmId,
      userId: req.user!.userId,
      action: 'audit_report_generated',
      entityType: 'compliance_scan_result',
      entityId: report.id,
    });

    res.status(201).json(serializeScanResult(report));
  } catch (err) {
    res.status(500).json({ error: true, message: 'Failed to generate audit report' });
  }
});

export default router;
